quixstreams/app.py

```diff
@@ -76,6 +86,288 @@ def __call__(
+            if self._stop_checker():
+                self.stop()
+
+    def start_runner(self):
+        """
+        Called as part of Application.run(), mostly to initialize self.running.
+        """
+        self.running = True
+        if self._is_first_run:
+            self._primary_topics = [t for t in self._topic_manager.topics]
+            self._repartition_topics = [
+                t for t in self._topic_manager.repartition_topics
+            ]
+            self._is_first_run = False
+
+    def stop(self):
+        """
+        Called when Application is stopped, or self._stop_checker condition is met.
+        Resets all values required for re-running.
+        """
+        self.running = False
+        self.last_consumed_tp = None
+        self._has_stop_checker = False
+        self._stop_checker = None
+
+        self._start_time = 0.0
+        self._timeout = 0.0
+        self._start_timeout_after_first_msg = False
+        self._needs_assigning = True
+        self._is_recovering = False
+
+        self._max_count = 0
+        self._count = 0
+        self._repartition_stop_points = {}
+
+    def handle_rebalance(self, recovery_required: bool):
+        """
+        This is the most common way the timeout start time will be set.
+        """
+        if self._timeout and self._needs_assigning:
+            self._needs_assigning = False
+            self._is_recovering = recovery_required
+            if not (self._start_timeout_after_first_msg or recovery_required):
+                self._set_timeout_start_time()
+
+    def handle_recovery(self):
+        """
+        Sets the timeout start time after a recovery occurs.
+        """
+        if self._timeout:
+            self._is_recovering = False
+            self._set_timeout_start_time()
+
+    def set_stop_condition(
+        self,
+        timeout: float = 0.0,
+        count: int = 0,
+        timeout_starts_on_first_message: bool = True,
+    ):
+        """
+        Called as part of app.run(); this handles the users optional stop conditions.
+        """
+        if not (timeout or count):
+            logger.info("APP RUN: preparing to run until errors or manually stopped...")
+            return
+
+        logger.info("APP RUN: preparing to run with stop conditions...")
+        if timeout:
+            if timeout < 0.0:
+                raise ValueError("run timeout must be >= 0.0")
+            self._start_timeout_after_first_msg = timeout_starts_on_first_message
+            if self._start_timeout_after_first_msg:
+                log_snippet = "(after consuming a non-recovery message)"
+            else:
+                log_snippet = "(after initial rebalance or recovery)"
+            logger.info(
+                f"APP RUN: timeout detected; "
+                f"Application will run for up to {timeout}s {log_snippet}"
+            )
+            self._stop_checker = self._at_timeout_func()
+        self._timeout = timeout
```

Comment:
I think it doesn't need to be re-set to 0.0
```suggestion
            self._timeout = timeout
```

Reply:
good catch; should have changed this after I tweaked them to no longer use None.